pkg/controller/podgc/gc_controller.go

```diff
@@ -41,61 +41,88 @@ const (
-	threshold      int
+	kubeClient clientset.Interface
+	podStore   cache.StoreToPodLister
+	nodeStore  cache.StoreToNodeLister
+
+	podController  cache.ControllerInterface
+	nodeController cache.ControllerInterface
+
+	deletePod              func(namespace, name string) error
+	terminatedPodThreshold int
 }
 
-func New(kubeClient clientset.Interface, resyncPeriod controller.ResyncPeriodFunc, threshold int) *PodGCController {
+func NewFromInformer(kubeClient clientset.Interface, podInformer cache.SharedIndexInformer, terminatedPodThreshold int) *PodGCController {
 	if kubeClient != nil && kubeClient.Core().GetRESTClient().GetRateLimiter() != nil {
 		metrics.RegisterMetricAndTrackRateLimiterUsage("gc_controller", kubeClient.Core().GetRESTClient().GetRateLimiter())
 	}
 	gcc := &PodGCController{
-		kubeClient: kubeClient,
-		threshold:  threshold,
+		kubeClient:             kubeClient,
+		terminatedPodThreshold: terminatedPodThreshold,
 		deletePod: func(namespace, name string) error {
 			return kubeClient.Core().Pods(namespace).Delete(name, api.NewDeleteOptions(0))
 		},
 	}
 
-	terminatedSelector := fields.ParseSelectorOrDie("status.phase!=" + string(api.PodPending) + ",status.phase!=" + string(api.PodRunning) + ",status.phase!=" + string(api.PodUnknown))
+	gcc.podStore.Indexer = podInformer.GetIndexer()
+	gcc.podController = podInformer.GetController()
 
-	gcc.podStore.Indexer, gcc.podStoreSyncer = cache.NewIndexerInformer(
+	gcc.nodeStore.Store, gcc.nodeController = cache.NewInformer(
 		&cache.ListWatch{
 			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
-				options.FieldSelector = terminatedSelector
-				return gcc.kubeClient.Core().Pods(api.NamespaceAll).List(options)
+				return gcc.kubeClient.Core().Nodes().List(options)
 			},
 			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
-				options.FieldSelector = terminatedSelector
-				return gcc.kubeClient.Core().Pods(api.NamespaceAll).Watch(options)
+				return gcc.kubeClient.Core().Nodes().Watch(options)
 			},
 		},
-		&api.Pod{},
-		resyncPeriod(),
+		&api.Node{},
+		controller.NoResyncPeriodFunc(),
 		cache.ResourceEventHandlerFuncs{},
-		// We don't need to build a index for podStore here actually, but build one for consistency.
-		// It will ensure that if people start making use of the podStore in more specific ways,
-		// they'll get the benefits they expect. It will also reserve the name for future refactorings.
-		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
 	)
+
 	return gcc
 }
 
+func NewFromClient(
+	kubeClient clientset.Interface,
+	terminatedPodThreshold int,
+) *PodGCController {
+	podInformer := informers.NewPodInformer(kubeClient, controller.NoResyncPeriodFunc())
+	return NewFromInformer(kubeClient, podInformer, terminatedPodThreshold)
+}
+
 func (gcc *PodGCController) Run(stop <-chan struct{}) {
-	go gcc.podStoreSyncer.Run(stop)
+	go gcc.podController.Run(stop)
```

Comment:
This won't work. If you are using SharedInformer, you shouldn't call Run for it.
Please change it to the pattern of "internalPodInformer" - an example of it you can find in ReplicationController:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replication/replication_controller.go#L88